sympy/matrices/common.py

```diff
@@ -1208,6 +1208,85 @@ def entry(i, j):
             return kls.zero
         return kls._new(size, size, entry)
 
+    def upper_triangular(self, kth_diagonal=0):
+        """returns the elements on and above the kth diagonal of a matrix.
+        If k is not specified then simply returns upper-triangular portion
+        of a matrix
+
+        Examples
+        ========
+
+        >>> from sympy import ones
+        >>> A = ones(4)
+        >>> A.upper_triangular()
+        Matrix([
+        [1, 1, 1, 1],
+        [0, 1, 1, 1],
+        [0, 0, 1, 1],
+        [0, 0, 0, 1]])
+
+        >>> A.upper_triangular(2)
+        Matrix([
+        [0, 0, 1, 1],
+        [0, 0, 0, 1],
+        [0, 0, 0, 0],
+        [0, 0, 0, 0]])
+
+        >>> A.upper_triangular(-1)
+        Matrix([
+        [1, 1, 1, 1],
+        [1, 1, 1, 1],
+        [0, 1, 1, 1],
+        [0, 0, 1, 1]])
+
+        """
+
+        R = self.zeros(self.rows, self.cols)
+
+        for j in range(self.cols):
+            if j - kth_diagonal >= 0:
+                R[:min(self.rows, j - kth_diagonal + 1), j] = self[:min(self.rows, j - kth_diagonal + 1), j]
```

Comment:
Does this work with immutable matrices?

Reply:
No it doesn't, the issue occours because R is defined as follows

```python
R = self.zeros(self.rows, self.cols)
```
So R also becomes an immutable matrix, is there some other way to define R as a matrix of zeros

Reply:
My opinion is you shouldn’t use assignment operation. 
Although you may just create a template list to mutate the items,
It would not be difficult to think of implementing this in other ways than assignment.
It’s expected that the output matrix should match the self.

Reply:
I've just converted the matrix R to mutable using as_mutable. If there is some issue with the latest commit let me know